### PR TITLE
Inferior chain macro block verification

### PR DESCRIPTION
**Pull request by nibhar**
This PR adds the ability to verify a macro block proposal whose predecessor is no longer on the main chain (i.e it was rebranched by a skip block in its place). Prior to this the state verification would fail, as the state was not altered from main chain head to the actual predecessor of the macro block.

Macro blocks with a non main chain predecessor are not invalid and thus their verification must succeed, given the micro blocks leading up to it.

### CODE REVIEW

File: validator/src/lib.rs [hunk 1]
```diff
@@ -5,5 +5,5 @@ pub mod aggregation;
 mod jail;
 mod r#macro;
```
Comment (paberr) [inline validator/src/lib.rs:6]: Is this a typo?

Comment (nibhar) [reply validator/src/lib.rs:6]: Not a typo. 

It is a raw identifier, as `macro` is a keyword. We could rename the file to not having to use it here, but it was not part of this PR.

File: blockchain/src/blockchain/rebranch_utils.rs [hunk 2]
```diff
@@ -0,0 +1,261 @@
...
+                    let remove_chain = vec![block.clone()]
+                        .into_iter()
+                        .chain(target_chain_iter.cloned())
+                        .collect();
+                    return Err(remove_chain);
+                }
+            }
+        }
+
+        // Unset onMainChain flag / mainChainSuccessor on the current main chain up to (excluding) the common ancestor.
```
Comment (hrxi) [inline blockchain/src/blockchain/rebranch_utils.rs]:
```suggestion
        // Unset on_main_chain flag / main_chain_successor on the current main chain up to (excluding) the common ancestor.
```

File: blockchain/src/blockchain/rebranch_utils.rs [hunk 3]
```diff
@@ -0,0 +1,261 @@
...
+        for reverted_block in revert_chain.iter_mut() {
+            reverted_block.1.on_main_chain = false;
+            reverted_block.1.main_chain_successor = None;
+
+            self.chain_store
+                .put_chain_info(write_txn, &reverted_block.0, &reverted_block.1, false);
+        }
+
+        // Update the mainChainSuccessor of the common ancestor block.
```
Comment (hrxi) [inline blockchain/src/blockchain/rebranch_utils.rs]:
```suggestion
        // Update the main_chain_successor of the common ancestor block.
```

File: blockchain/src/blockchain/rebranch_utils.rs [hunk 4]
```diff
@@ -0,0 +1,261 @@
...
+        ancestor.1.main_chain_successor = Some(target_chain.last().unwrap().0.clone());
+        self.chain_store
+            .put_chain_info(write_txn, &ancestor.0, &ancestor.1, false);
+
+        // Set onMainChain flag / mainChainSuccessor on the fork.
```
Comment (hrxi) [inline blockchain/src/blockchain/rebranch_utils.rs]:
```suggestion
        // Set on_main_chain flag / main_chain_successor on the fork.
```

File: blockchain/src/blockchain/verify.rs [hunk 5]
```diff
@@ -296,4 +301,237 @@ impl Blockchain {
...
+        round: u32,
+        valid_round: Option<u32>,
+        signed_data: Vec<u8>,
+        signature: &Signature,
+        signature_only: bool,
+    ) -> Result<MacroBody, PushError> {
+        // If the accounts tree is not complete nothing can be done
+        if !self.accounts_complete() {
+            // Maybe introduce a proper error here. Is this even needed?
+            return Err(PushError::MissingAccountsTrieDiff);
```
Comment (hrxi) [inline blockchain/src/blockchain/verify.rs]:
```suggestion
            return Err(PushError::IncompleteAccountsTrie);
```

File: blockchain/src/blockchain/verify.rs [hunk 6]
```diff
@@ -296,4 +301,237 @@ impl Blockchain {
...
+        proposed_block: MacroBlock,
+        round: u32,
+        valid_round: Option<u32>,
+        signed_data: Vec<u8>,
+        signature: &Signature,
+        signature_only: bool,
+    ) -> Result<MacroBody, PushError> {
+        // If the accounts tree is not complete nothing can be done
+        if !self.accounts_complete() {
+            // Maybe introduce a proper error here. Is this even needed?
```
Comment (hrxi) [inline blockchain/src/blockchain/verify.rs:329]: Theoretically, the validator isn't active while the trie isn't complete, so this should not happen?

A validator will just vote none if this function returns an error, but that's not a problem in terms of the protocol, right?

File: validator/src/tendermint.rs [hunk 7]
```diff
@@ -330,134 +329,40 @@ where
...
-        }
-
-        // Abort if the blockchain is not in the correct state anymore.
-        let blockchain = self.blockchain.read();
-        if !blockchain.accounts_complete() {
-            return Err(ProposalError::Other);
-        }
+        // verify_macro_block_proposal could create and commit a write transaction, thus take an
+        // upgradable read lock on blockchain.
+        let blockchain = self.blockchain.upgradable_read();
```
Comment (hrxi) [inline validator/src/tendermint.rs]: This can go below so that the lock is only held for the minimum time.

File: blockchain/src/blockchain/rebranch_utils.rs [hunk 8]
```diff
@@ -0,0 +1,261 @@
...
+        // Revert to common ancestor is done.
+
+        // Next, push each block of the target chain.
+
+        // Pushing must happen in reverse.
+        let mut target_chain_iter = target_chain.iter().rev();
+
+        while let Some(block) = target_chain_iter.next() {
+            // Collect logs for the upcoming push.
+            let mut block_logger = BlockLogger::new_applied(
```
Comment (hrxi) [inline blockchain/src/blockchain/rebranch_utils.rs:184]: Sneaky change, could maybe go into another commit or in the commit message.

File: blockchain/src/blockchain/verify.rs [hunk 9]
```diff
@@ -296,4 +301,237 @@ impl Blockchain {
...
+            .map_err(PushError::BlockchainError)?
+            .header();
+
+        // Get the validator for this round.
+        let mut proposer = self
+            .get_proposer_at(
+                proposed_block.block_number(),
+                round,
+                prev_header.seed().entropy(),
+                None,
```
Comment (hrxi) [inline blockchain/src/blockchain/verify.rs]: Nit, but not required: We pass `None` for the transaction here and below, if we created a read transaction here, we wouldn't need to open two to get two proposers (also existed before this PR).

File: validator/src/tendermint.rs [hunk 10]
```diff
@@ -330,134 +329,40 @@ where
...
+                log::debug!(?error, ?proposal, "Proposal verification failed",);
+                // Special case for invalid signatures, as that means the proposer is not the signer.
+                // It is needed as the proposer if he produces a faulty block might be subjected to
+                // some sort of punishment or mitigation whereas with an invalid signature it would be
+                // the one relaying the message, who would potentially be subjected to protective measures.
+                if error == PushError::InvalidBlock(BlockError::InvalidJustification) {
+                    ProposalError::InvalidSignature
+                } else {
+                    ProposalError::InvalidProposal
+                }
```
Comment (hrxi) [inline validator/src/tendermint.rs:364]: Just an idea, not mandatory: Maybe `PushError` is the wrong error type for `verify_macro_block_proposal`, maybe it could return a `ProposalError` directly?

**Review by hrxi — COMMENTED** [review]
Looks very good, I only had a couple of nits that aren't required to be fixed and two questions.

File: blockchain/src/blockchain/verify.rs (hunk 6 above)
Comment (nibhar) [reply blockchain/src/blockchain/verify.rs:329]: Correct. Deeming a proposal invalid in the case where you cannot verify its validity is the only safe option. I do not quite know about wether or not a validator with an incomplete tree can consider itself active, as i have not worked on it. I would assume, it is not fit for validating with an incomplete tree, but i would have to check with who worked on that.
This might thus be `unreachable!()`.

File: validator/src/tendermint.rs (hunk 10 above)
Comment (nibhar) [reply validator/src/tendermint.rs:364]: `ProposalError` is a fairly unspecific error type. I chose to only convert it here such that it can be logged before the conversion to hopefully help with debugging and for the, albeit highly unlikely chance, it would potentially see use in a different location. 

`ProposalError` is also an export of `nimiq-tendermint` which `nimiq-blockchain` currently only uses as a dev dependency, which I wanted to maintain.

Personally I would rather keep the `PushError` as it also is in line with the remaining functions present in `blockchain/src/blockchain/verify.rs` but changing it is not a big deal if it is considered better form.

File: blockchain/src/blockchain/verify.rs (hunk 9 above)
Comment (nibhar) [reply blockchain/src/blockchain/verify.rs]: I will also include the same `txn` in the query for the preceding block.

File: blockchain/src/blockchain/rebranch_utils.rs (hunk 8 above)
Comment (nibhar) [reply blockchain/src/blockchain/rebranch_utils.rs:184]: Done

**Comment by hrxi** [comment]
Rebased to get it signed and pushed to master: https://github.com/nimiq/core-rs-albatross/commit/08e0d65a344585ef0af1b27dd84a71f306e3f8d0.
